[WIP] Experimental DVD chunking

mmastrac opened the pull request:
Experimental build: breaks DVD reads up into chunks to (hopefully) fix games that rely on data being there shortly after issuing a read, while not breaking games that expect data later in the buffer to the there until the read is nearly complete.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/3701)

<!-- Reviewable:end -->

JMC47 commented:
@dolphin-emu-bot rebuild

mmastrac commented:
This is currently WIP: the actual timing is off and I'm pretty sure I don't have the buffered reads nailed just yet. I am going to tweak this until it better matches the output from the hwtest. This should be useful for testing games that have trouble running with or without SUDTR, however.

JosJuice commented:
I did a quick look through this without reading the buffer logic in detail. It would be simpler and more efficient to submit the parts of the read all at once to a queue in DVDThread instead of doing it with callbacks and additional state in DVDInterface, but too bad, we don't have any queue in DVDThread right now... :) I suppose this PR can continue using its current approach, and if I add a queue to DVDThread in a later PR, this can be changed over then. Though if you want to make such a queue on your own, feel free to.

~~but does it fix Sonic Riders??~~

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 35 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBvg1zmgcXBNTzUxXn9:-KBvg1zmgcXBNTzUxXnA:884418389) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7aff6f83eda8c2a197534bd02864d8d120206382/Source/Core/Core/HW/DVDInterface.cpp#L35)):*
I don't think there's supposed to be a ß here.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 42 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBvg5QleKFETMtY3sgo:-KBvg5QmKHeVTr2wcz2U:1536327901) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7aff6f83eda8c2a197534bd02864d8d120206382/Source/Core/Core/HW/DVDInterface.cpp#L42)):*
This doesn't match DVD terminology. A sector contains 2 KiB of user data and an ECC block contains 32 KiB of user data.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 269 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBvgQxtqS1XDVFjSmGO:-KBvgQxtqS1XDVFjSmGP:907284465) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7aff6f83eda8c2a197534bd02864d8d120206382/Source/Core/Core/HW/DVDInterface.cpp#L269)):*
Could we have a struct that contains all of these (other than index, maybe)? Defining these five as separate variables both in here and in DVDThread is a bit redundant.

Also, you need to include all the new variables (except s_finish_read) in DoState and update the state version number in State.cpp.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 332 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBvlkaQiRB1QUPw8htp:-KBvlkaRhx7rDjc6urzC:-1129992896) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7aff6f83eda8c2a197534bd02864d8d120206382/Source/Core/Core/HW/DVDInterface.cpp#L332)):*
Should you compensate for `cyclesLate` by making the delay before the next scheduled event shorter?

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1295 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBvjd2tq3qQKUTYCGKL:-KBvjd2uAXOQ4pqJ474K:1035228368) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a64dd3659ad9b8bf9c6297124f0496d3f1aa0df2/Source/Core/Core/HW/DVDInterface.cpp#L1295)):*
I would like to keep a comment somewhere saying that 50 ms is a rough approximation. It's probably the biggest loading time inaccuracy we have right now.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1317 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBvihBr5FXxt7Nxc-_D:-KBvihBr5FXxt7Nxc-_E:-503813900) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7aff6f83eda8c2a197534bd02864d8d120206382/Source/Core/Core/HW/DVDInterface.cpp#L1317)):*
Won't this multiplication overflow u32, or does the result get promoted to u64 somehow? You should be able to do `SystemTimers::GetTicksPerSecond() / 1000000 * COMMAND_LATENCY_US` instead without getting rounding errors, because `GetTicksPerSecond()` can only return 486000000 or 729000000.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-:-KBvk13vcyxvHhcs4ann:851712866)*
<!-- Sent from Reviewable.io -->

JMC47 commented:
Fixes Sonic Riders.  Please merge.  Also the DVD timings seem WAY better.

JosJuice commented:
@JMC47: This should wait until after 5.0.

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 332 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBvlkaQiRB1QUPw8htp:-KBw9ev0S9vEVpOJ54Qg:-811708744) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7aff6f83eda8c2a197534bd02864d8d120206382/Source/Core/Core/HW/DVDInterface.cpp#L332)):*
Actually, nevermind. Because of how DVDThread works, `cyclesLate` will always be 0 here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-:-KBw9tJyX7NVg-1cy5O5:1236627599)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 671 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBxHFIXHwYKHVWj2rRa:-KBxHFIXHwYKHVWj2rRb:194143214) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a64dd3659ad9b8bf9c6297124f0496d3f1aa0df2/Source/Core/Core/HW/DVDInterface.cpp#L671)):*
Would we need to remove the fast disc speed option? It can be kept as an enhancement for people who want faster loading times while not caring about accuracy.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 671 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBxHFIXHwYKHVWj2rRa:-KBxQdZ1UlpcVpPd7kOs:-1786390141) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a64dd3659ad9b8bf9c6297124f0496d3f1aa0df2/Source/Core/Core/HW/DVDInterface.cpp#L671)):*
I don't see why we would need to remove it. It's currently re-implemented further down in the file.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701)*
<!-- Sent from Reviewable.io -->

mmastrac commented:
The patch is still WIP (state stuff has been omitted to make testing easier), but here's a status update:

 - I've confirmed that the streaming read time accurately match the hardware test. 
 - The DVD interface is now simulating seek times, but those still need a bit more tuning to match the hardware test.
 - The buffer code appears to be functioning properly, but needs more extensive validation (and potentially another hardware test or two) before I'd call it done.

JMC47 commented:
Fixes some Scooby Doo game too.

BhaaLseN commented:
Code-style looks ok-ish (other than the things mentioned), love the comments tho! Details like that need to be documented, or we'll all stumble around in the dark if anything ever comes up (altho, this isn't necessarily just limited to DVDInterface).

<details><summary><i>Previously, JMC47 wrote…</i></summary>
> Fixes some Scooby Doo game too.
</details>


---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1520 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KIHvkRRIf9AAnxaadlm:-KIHvkRRIf9AAnxaadln:898463742) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9cd8e140b550c5daa762189c1d46c3b148e92246/Source/Core/Core/HW/DVDInterface.cpp#L1520)):*
> ```cpp
> 	return std::sqrt((double)offset / WII_DISC_LAYER_SIZE
> 		             * (WII_DVD_OUTER_RADIUS*WII_DVD_OUTER_RADIUS-DVD_INNER_RADIUS*DVD_INNER_RADIUS)
> 		             + DVD_INNER_RADIUS*DVD_INNER_RADIUS);
> ```

Some whitespace around the operators please?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-:-KIHw1HpVMYtyB5WMU7j:1483667374)*
<!-- Sent from Reviewable.io -->

mmastrac commented:
What's left for this patch?

- [ ] Run the hardware tests for dual layer DVDs
- [ ] Run the hardware tests for gamecube disks
- [ ] Savestate work
- [ ] Tune the constants to match real-world data

JosJuice commented:
For dual-layer reading, you may want to borrow my code from PR #1900. I made the same assumption as you did about the second layer, but tueidj said that layer 2 actually goes backwards compared to layer 1, so I changed the code.

mmastrac commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 42 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBvg5QleKFETMtY3sgo:-KC1F7zV2v3XvheMko6R:b-3mn44z) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7aff6f83eda8c2a197534bd02864d8d120206382/Source/Core/Core/HW/DVDInterface.cpp#L42)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> This doesn't match DVD terminology. A sector contains 2 KiB of user data and an ECC block contains 32 KiB of user data.
</details>
Agreed. I'll call this DVD_ECC_BLOCK_SIZE

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1295 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBvjd2tq3qQKUTYCGKL:-KOggpUIq9AAdAu1f513:b-wnyr6t) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a64dd3659ad9b8bf9c6297124f0496d3f1aa0df2/Source/Core/Core/HW/DVDInterface.cpp#L1295)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> I would like to keep a comment somewhere saying that 50 ms is a rough approximation. It's probably the biggest loading time inaccuracy we have right now.
</details>
Seek times are now roughly emulated based on a hwtest model

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1317 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBvihBr5FXxt7Nxc-_D:-KC1FcW79l5Qwn6MY22J:bcrv367) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7aff6f83eda8c2a197534bd02864d8d120206382/Source/Core/Core/HW/DVDInterface.cpp#L1317)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Won't this multiplication overflow u32, or does the result get promoted to u64 somehow? You should be able to do `SystemTimers::GetTicksPerSecond() / 1000000 * COMMAND_LATENCY_US` instead without getting rounding errors, because `GetTicksPerSecond()` can only return 486000000 or 729000000.
</details>
Good catch: you are right - all of these are u32s, which means there is no widening in the expression. I can rework this.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1520 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KIHvkRRIf9AAnxaadlm:-KOggvQZGjJQ2wtwwOft:b9w2ehi) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9cd8e140b550c5daa762189c1d46c3b148e92246/Source/Core/Core/HW/DVDInterface.cpp#L1520)):*
<details><summary><i>Previously, BhaaLseN (BhaaL) wrote…</i></summary>
> Some whitespace around the operators please?
</details>
Fixed via clang-format

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701)*
<!-- Sent from Reviewable.io -->

JMC47 commented:
Going to retest a few bugs on this.

mmastrac commented:
@JMC47, stuff to test:

- Sonic Riders
- Anything GameCube 
- Any disc with two layers (eg: Last Story)
- Save state during disc access

mmastrac commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 269 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KBvgQxtqS1XDVFjSmGO:-KOgkTUungs0N3kADWi9:bj6bhek) ([raw file](https://github.com/dolphin-emu/dolphin/blob/7aff6f83eda8c2a197534bd02864d8d120206382/Source/Core/Core/HW/DVDInterface.cpp#L269)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Could we have a struct that contains all of these (other than index, maybe)? Defining these five as separate variables both in here and in DVDThread is a bit redundant.
> 
> Also, you need to include all the new variables (except s_finish_read) in DoState and update the state version number in State.cpp.
</details>
Updated the state code. Honestly, not sure if it's worth creating a struct as I've removed some of the stuff from DVDThread that's no longer useful.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701)*
<!-- Sent from Reviewable.io -->

mmastrac commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1520 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KIHvkRRIf9AAnxaadlm:-KOgl84OboIZ45gadiMn:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/9cd8e140b550c5daa762189c1d46c3b148e92246/Source/Core/Core/HW/DVDInterface.cpp#L1520)):*
<details><summary><i>Previously, mmastrac (Matt Mastracci) wrote…</i></summary>
> Fixed via clang-format
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701)*
<!-- Sent from Reviewable.io -->

JMC47 commented:
You update it so often that I can't quite get to testing it before the updates :D.  I like updates though

mmastrac commented:
@JMC47 all done now. I wanted to clean up the commit message for eventual merge.

JMC47 commented:
Arc Rise Fantasia works

JMC47 commented:
Tested Brawl, The Last Story, Xenoblade and Raving Rabbids Party Collection.  Those are my duallayer games.

JMC47 commented:
Did a savestate/loadstate during the fateful crash in Sonic Riders and it proceeded normally.

mmastrac commented:
@JosJuice Could use a final once-over from you. Note that I stole your dual-layer offset code from #1900. 😄

JMC47 commented:
Fixes F1 2002's epic load screen,

JosJuice commented:
Here's a preliminary review (I haven't taken the time to understand the buffer logic yet). It looks good in general, but the way DVDThread is being used is both bad for the performance and a bit ugly. We could really use a queue in DVDThread like I described before. I will try to make a PR for that in the upcoming days.

---

Reviewed 1 of 4 files at r5.
Review status: 1 of 4 files reviewed at latest revision, 16 unresolved discussions.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 53 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjAjau6D6SByjdH-fU:-KOjAjawqImE1paNWpu3:bond0rw) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L53)):*
> ```cpp
> 
> // The size of a Wii disc layer in bytes (2294912 sectors per layer)
> constexpr u64 WII_DISC_LAYER_SIZE = 2294912 * DVD_SECTOR_SIZE;
> ```

This size is only correct for the first layer, not the second. The code is fine, but you might want to note this in the comment.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 55 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjBU5vWU-z6el5xJfi:-KOjBU5vWU-z6el5xJfj:b-4hv8rv) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L55)):*
> ```cpp
> constexpr u64 WII_DISC_LAYER_SIZE = 2294912 * DVD_SECTOR_SIZE;
> 
> // 24mm
> ```

If I'm allowed to nitpick, there should be a space between the number and unit here and elsewhere in the PR.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 63 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjB39KNIyJnQQy_chN:-KOjB39L3-0E7Ki6O0rd:box4xki) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L63)):*
> ```cpp
> 
> // Approximate read speeds at the inner and outer locations of Wii and GC
> // discs. These speeds are approximations of speeds measured on real Wiis.
> ```

Please include the unit (bytes per second) in the comment.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 317 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOk9Cw55mqNIGv5ZQlL:-KOk9Cw6jPlFc4Oin8S2:b-qrepc0) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L317)):*
> ```cpp
>   if (offset == 0)
>   {
>     if (!DVDThread::CompleteRead())
> ```

Forcing DVDThread to finish the whole read once we need the first ECC block somewhat negates the point of having a separate thread for reading the disc. Because of the unnecessary performance impact, I'm not fine with merging the PR if it will work like this.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 318 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOkDV9_y-SWCiVpfyKQ:-KOkDV9a4qXqtvI-TSaB:b-cac58b) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L318)):*
> ```cpp
>   {
>     if (!DVDThread::CompleteRead())
>       return;
> ```

There are two problems here.

1. If a command has been chunked into only one read, returning here will skip calling FinishExecutingCommand (and DVDThread::Cleanup). Calling FinishExecutingCommand with INT_TCINT when a read fails doesn't make much sense anyway, but I'm assuming this change in behavior isn't intentional. Actually signaling that the read failed would be optimal, but that's out of scope for this PR.

2. If a command has been chunked into more than one read and CompleteRead fails, DMA will be called for all scheduled events other than the first one, triggering debug asserts and other bad things.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 707 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjqn2le_ggnwy6UvdI:-KOjqn2mNRDoKV87Sp1J:b-fowmjq) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L707)):*
> ```cpp
> {
>   DIInterruptType interrupt_type = INT_TCINT;
>   u64 ticks_until_completion = SystemTimers::GetTicksPerSecond() / 15000;
> ```

This doesn't need to be a variable anymore. You can just use a constant down in the call to ScheduleEvent. You should probably use COMMAND_LATENCY_US instead of this arbitrary 1 / 15000.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1338 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjum0cAG_bjGeslFzn:-KOjum0dyLuM9df61pgd:bs9i5yp) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1338)):*
> ```cpp
>   //
>   // Metroid Prime is a good example of a game that's sensitive to all of
>   // these details; if there isn't enough latency in the right places, doors
> ```

Maybe reword "all of these details", since the text above doesn't mention seek timing anymore.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1539 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOk1rXC5LDMJ0jxCz1q:-KOk1rXDocyivX6MtnD1:b-9fwdc0) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1539)):*
> ```cpp
> //
> // Where:
> //   P = the offset's linear position, as offset/density
> ```

I don't understand what P is :(
From solving the `L = pi*(r.outer^2-r.inner^2)/pitch` equation, I could understand that L is the same as P, but I don't know how density ties into that...

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1548 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOk33oYP9AIris202uv:-KOk33oYP9AIris202uw:b-rmauyt) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1548)):*
> ```cpp
> // given byte offset as a drastically simplified:
> //
> // r = sqrt(offset/total_bytes*(r.outer^2-r.inner^2) + r.inner^2)
> ```

Reminder to myself to go back later and actually understand what this is doing

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1553 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOk--br8IQZ-DeHNsBS:-KOk--bsd-A_dSI65tA7:bm2xck8) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1553)):*
> ```cpp
>   // Assumption: the layout on the second disc layer is opposite of the first, ie
>   // layer 2 starts where layer 1 ends and goes backwards.
>   if (offset > WII_DISC_LAYER_SIZE) {
> ```

`>=` makes more sense to me here (though I suppose the result would be the same regardless, so it's not important)

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1555 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjzU8g6qE8VXXxyfGq:-KOjzU8iezQjnU2Wm6Yv:b-feai4g) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1555)):*
> ```cpp
>   if (offset > WII_DISC_LAYER_SIZE) {
>     offset = WII_DISC_LAYER_SIZE * 2 - offset;
>   }
> ```

Preferably remove the { and }. If not, the { needs to be on a new line.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1592 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjsbtClS4AL4uGaFl0:-KOjsbtDtKEmqonogaqh:b-pot49d) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1592)):*
> ```cpp
> // Returns the number of ticks it takes to read an amount of data from a disc,
> // ignoring factors such as seek times. This is the streaming rate of the
> // drive and varies between ~3-8MiB/s. Note that there is technically a DMA
> ```

It starts at roughly 2 MiB/s for GC.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1593 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjsSsjZDGkaArMmxvw:-KOjsSskOc8p1mGBArPj:bxxqqi) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1593)):*
> ```cpp
> // ignoring factors such as seek times. This is the streaming rate of the
> // drive and varies between ~3-8MiB/s. Note that there is technically a DMA
> // delay on top of this, but we model that as part of this read time.ß
> ```

There's a stray ß again.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1596 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjszTKtH14fMNMktzy:-KOjszTLa4TguiGYQtp6:b-tnks9l) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1596)):*
> ```cpp
> u64 CalculateRawDiscReadTime(u64 offset, u32 length)
> {
>   // The Wii is a CAV drive and the data has a constant pit length
> ```

"GameCube and Wii" for clarity, not just "Wii".

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1599 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjwd55hbPEiEXcJSSN:-KOjwd56T-oNsUZssyvk:bg1sryl) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1599)):*
> ```cpp
>   // regardless of location on disc. This means we can linearly interpolate
>   // speed from the inner to outer radius. This matches a hardware test.
>   double physical_offset = CalculatePhysicalDiscPosition(offset + length / 2);
> ```

Isn't it more accurate to do `(CalculatePhysicalDiscPosition(offset) + CalculatePhysicalDiscPosition(offset + length)) / 2;`? (Though I suppose that has small problems for reads that span two layers...)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-:-KOk6VpbJku38xYLaMIv:b-nsskkp)*
<!-- Sent from Reviewable.io -->

MaJoRoesch commented:
@lioncash do you want to do a code style review of this PR?

mmastrac commented:
Review status: 1 of 4 files reviewed at latest revision, 16 unresolved discussions.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 317 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOk9Cw55mqNIGv5ZQlL:-KOkJR_WE29gQyHj8Vxo:b2nilzv) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L317)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Forcing DVDThread to finish the whole read once we need the first ECC block somewhat negates the point of having a separate thread for reading the disc. Because of the unnecessary performance impact, I'm not fine with merging the PR if it will work like this.
</details>
It's not as bad as you think, but I can understand what you are thinking.

We run the whole DVD read asynchronously, starting from when the command is submitted_ and_ while the delay for the command *PLUS* initial seek *PLUS* initial read chunk happens (always >= DVD ECC block size). I think this is probably not a big deal as the hardware we are running on is *much* faster than the Wii's drive. We only have to be able to do a native read of the whole buffer in the about 1/4 the time of the Wii's drive to fit into that initial chunk. 

The vast majority of users will be running games from SD card, HDD or SSD, all of which will have way better performance characteristics than the Wii drive. The worst case for this will be users running their games off DVD drives, and even those drives will be significantly ahead of the Wii's technology.

I could investigate chunking the native read up as well, but that's going to add a lot of complexity to this.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 318 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOkDV9_y-SWCiVpfyKQ:-KOkM7UKD5-mCOr3mW5w:barud3i) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L318)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> There are two problems here.
> 
> 1. If a command has been chunked into only one read, returning here will skip calling FinishExecutingCommand (and DVDThread::Cleanup). Calling FinishExecutingCommand with INT_TCINT when a read fails doesn't make much sense anyway, but I'm assuming this change in behavior isn't intentional. Actually signaling that the read failed would be optimal, but that's out of scope for this PR.
> 
> 2. If a command has been chunked into more than one read and CompleteRead fails, DMA will be called for all scheduled events other than the first one, triggering debug asserts and other bad things.
</details>
What's the correct way to signal that a read failed? I agree that calling DMA for all the other scheduled events is bad.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1539 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOk1rXC5LDMJ0jxCz1q:-KOkIp1iLSswLGY5d2Qq:b2xhzkv) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1539)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> I don't understand what P is :(
> From solving the `L = pi*(r.outer^2-r.inner^2)/pitch` equation, I could understand that L is the same as P, but I don't know how density ties into that...
</details>
That was a typo - `P` should be `L`.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1592 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjsbtClS4AL4uGaFl0:-KOkJ48TkTA-dcS7LNGJ:bs7rzi8) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1592)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> It starts at roughly 2 MiB/s for GC.
</details>
Ah yeah, we have that in the constant but it needs to be mentioned here.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1599 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjwd55hbPEiEXcJSSN:-KOkJCv0NoU5q98LWQwV:b-hrt21n) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1599)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Isn't it more accurate to do `(CalculatePhysicalDiscPosition(offset) + CalculatePhysicalDiscPosition(offset + length)) / 2;`? (Though I suppose that has small problems for reads that span two layers...)
</details>
It's such a small rounding error in the grand scheme of things that it doesn't matter. In fact, I was considering just taking the physical offset at the start since that yields pretty much identical results.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 1 of 4 files reviewed at latest revision, 16 unresolved discussions.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 317 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOk9Cw55mqNIGv5ZQlL:-KOkUrS3gkp3tRk1RQbk:b-lm2eh2) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L317)):*
<details><summary><i>Previously, mmastrac (Matt Mastracci) wrote…</i></summary>
> It's not as bad as you think, but I can understand what you are thinking.
> 
> We run the whole DVD read asynchronously, starting from when the command is submitted_ and_ while the delay for the command *PLUS* initial seek *PLUS* initial read chunk happens (always >= DVD ECC block size). I think this is probably not a big deal as the hardware we are running on is *much* faster than the Wii's drive. We only have to be able to do a native read of the whole buffer in the about 1/4 the time of the Wii's drive to fit into that initial chunk. 
> 
> The vast majority of users will be running games from SD card, HDD or SSD, all of which will have way better performance characteristics than the Wii drive. The worst case for this will be users running their games off DVD drives, and even those drives will be significantly ahead of the Wii's technology.
> 
> I could investigate chunking the native read up as well, but that's going to add a lot of complexity to this.
> 
</details>
I agree, it's not as bad as it can sound. Users are unlikely to notice it. (They didn't even really notice the slowdown from only having blocking disc I/O..) Still, I don't think it's a good idea to have the complexity of a separate thread without making full use of it.

I'm going to work on adding a DVDThread queue, which will make chunking the native read simple afterwards. Let's hope I don't take too long...

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 318 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOkDV9_y-SWCiVpfyKQ:-KOkWSnHESnnTT6GnrJg:b-lj9flb) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L318)):*
<details><summary><i>Previously, mmastrac (Matt Mastracci) wrote…</i></summary>
> What's the correct way to signal that a read failed? I agree that calling DMA for all the other scheduled events is bad.
</details>
I believe you have to use INT_DEINT instead of INT_TCINT and set s_error_code as described by 5.7.3.5.1 of YAGCD.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 1 of 4 files reviewed at latest revision, 16 unresolved discussions.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 318 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOkDV9_y-SWCiVpfyKQ:-KOkXkGaWz06zZ1UVTVo:b7d7msx) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L318)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> I believe you have to use INT_DEINT instead of INT_TCINT and set s_error_code as described by 5.7.3.5.1 of YAGCD.
</details>
Signaling failure will probably break things for the reporter of https://bugs.dolphin-emu.org/issues/7930, but since their ISO actually is slightly invalid, I'm not going to treat that as a problem.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1599 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjwd55hbPEiEXcJSSN:-KOkXWQSeoi4q7Q5lcXd:bw9u315) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1599)):*
<details><summary><i>Previously, mmastrac (Matt Mastracci) wrote…</i></summary>
> It's such a small rounding error in the grand scheme of things that it doesn't matter. In fact, I was considering just taking the physical offset at the start since that yields pretty much identical results.
</details>
I'm fine with that if you leave a small comment about it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701)*
<!-- Sent from Reviewable.io -->

lioncash commented:
@MaJoRoesch Sure, I can do that after work.

mmastrac commented:
Review status: 1 of 4 files reviewed at latest revision, 15 unresolved discussions.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 707 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjqn2le_ggnwy6UvdI:-KOlSwFRHLJXuZXyV022:bj2q8i1) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L707)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> This doesn't need to be a variable anymore. You can just use a constant down in the call to ScheduleEvent. You should probably use COMMAND_LATENCY_US instead of this arbitrary 1 / 15000.
</details>
I'll leave it at the current value until we write a hwtest for it, but I'll move it down to where it's used.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1555 \[r5\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOjzU8g6qE8VXXxyfGq:-KOlTSw-j84a0JauETmx:bfautws) ([raw file](https://github.com/dolphin-emu/dolphin/blob/04973c949d53cbe757ece2c593fe2178a09f50b9/Source/Core/Core/HW/DVDInterface.cpp#L1555)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Preferably remove the { and }. If not, the { needs to be on a new line.
</details>
Oops, yeah - formatting issue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 1 of 4 files reviewed at latest revision, 8 unresolved discussions.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 317 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOnGbSFcX1DR-p9T755:-KOnGbSFcX1DR-p9T756:bpq75yc) ([raw file](https://github.com/dolphin-emu/dolphin/blob/141bc9288c21bd52e43ab1c65f6ff9b9d7c276dc/Source/Core/Core/HW/DVDInterface.cpp#L317)):*
> ```cpp
>   bool complete = offset + length == s_current_read_length;
> 
>   if (!DVDThread::CompleteRead())
> ```

Calling CompleteRead once for each chunk seems suspicious. DEBUG_LOG will be triggered multiple times, and PanicAlertT will also be triggered multiple times if the read failed.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 325 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOnH92ltDfshoqd0Ocb:-KOnH92m0aPX9J4hTbSZ:b-qjsne) ([raw file](https://github.com/dolphin-emu/dolphin/blob/141bc9288c21bd52e43ab1c65f6ff9b9d7c276dc/Source/Core/Core/HW/DVDInterface.cpp#L325)):*
> ```cpp
>       DEBUG_LOG(DVDINTERFACE, "Read failed");
>       DVDThread::Cleanup();
>       FinishExecutingCommand(s_current_read_reply_to_ios, DVDInterface::INT_DEINT);
> ```

The error code needs to be set before calling FinishExecutingCommand with INT_DEINT.

I would prefer to keep this new error handling to another PR. This one is already full of various things.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701)*
<!-- Sent from Reviewable.io -->

lioncash commented:
Review status: 1 of 4 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1364 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOob5KJQn6r3ArirJ8I:-KOob5KK4G5yzSFwoReK:b-j2vi9v) ([raw file](https://github.com/dolphin-emu/dolphin/blob/141bc9288c21bd52e43ab1c65f6ff9b9d7c276dc/Source/Core/Core/HW/DVDInterface.cpp#L1364)):*
> ```cpp
>   u64 head_position;
> 
>   // Compute the start (inclusive) and end (exclusize) of the buffer. If we
> ```

`exclusive`

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1423 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOodzRv9g1CMtA6tkbn:-KOodzRwWGQ9XVjvyzI9:bm5dzd) ([raw file](https://github.com/dolphin-emu/dolphin/blob/141bc9288c21bd52e43ab1c65f6ff9b9d7c276dc/Source/Core/Core/HW/DVDInterface.cpp#L1423)):*
> ```cpp
>             buffer_start, buffer_end, buffer_end - buffer_start);
> 
>   u32 offset = 0;
> ```

You can likely eliminate the cast to `u64` down below, by just making this a `u64`.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1432 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOoZ1pG1XKWYOVTiFxA:-KOoZ1pHU_PoJ5HtmSI6:bp33adl) ([raw file](https://github.com/dolphin-emu/dolphin/blob/141bc9288c21bd52e43ab1c65f6ff9b9d7c276dc/Source/Core/Core/HW/DVDInterface.cpp#L1432)):*
> ```cpp
>   u64 ticks_until_completion = COMMAND_LATENCY_US * (SystemTimers::GetTicksPerSecond() / 1000000);
> 
>   int buffered_blocks = 0;
> ```

You may want to make this a`u32` considering they're used within a loop that uses a `u32` as the test condition.

---

*[Source/Core/Core/HW/DVDThread.cpp, line 120 \[r6\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KOohjUuu5BALAM9T7-k:-KOohjUviYHFt0fPknQx:b-c8oqex) ([raw file](https://github.com/dolphin-emu/dolphin/blob/141bc9288c21bd52e43ab1c65f6ff9b9d7c276dc/Source/Core/Core/HW/DVDThread.cpp#L120)):*
> ```cpp
> }
> 
> void DMA(u32 chunk_offset, u32 output_address, u32 length)
> ```

`length` is likely best suited as a `size_t`, as that's the type of the length parameter for `CopyToEmu`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
@mmastrac The queue PR is #4120, if you want to take a look at it. Add a `NONE` entry to the `ReplyType` enum, then you can submit all chunks of each read at once to DVDThread, using `NONE` as the `ReplyType` for all chunks other than the last.

RisingFog commented:
What's the status of this PR?

mmastrac commented:
@RisingFog I think there are a few comments left to address (and a rebase) but it's mostly ready to go.

hyperspeedgx commented:
You guys may want to check this: https://forums.dolphin-emu.org/Thread-metroid-prime-gcn-ntsc-no-boss-music?pid=423338#pid423338

I don't remember of that problem happening on my GC, so I suspect that the loads are so fast that it won't play the next BGM, that save room door needs to take more time to load like a real GC, if I remember well the 1st door always take more time to load on GC, that way the BGM issue in Metroid Prime might become fully fixed, except on the last battle vs Metroid Prime, even on the real GC if you skip the cutscene the BGM won't play. :)

JMC47 commented:
We'll see.  This needs to be rebased and there's another pull request I think this relies on before it'll be merged.  Regardless of if it fixes that issue, it fixes Sonic Riders, which as we all know is the top selling GameCube game with ~7 billion copies sold as of this date.  It also inspired other popular games, like Super Mario 64, Ratchet and Clank, The Legend of Zelda: A Link to the Past, and Sonic The Hedgehog 2.

JosJuice commented:
@mmastrac Any chance for a rebase soon? The DVD queue PR was merged a little while ago.

iwubcode commented:
@JosJuice  - silly question, if the original author doesn't rebase this, can someone else?

JosJuice commented:
Yes. If @mmastrac doesn't rebase it, the plan is that I will rebase it and submit it as a new PR. But I'm not sure when I'll have enough time to both finish up my review and do the rebase.

iwubcode commented:
@JosJuice - ok, please let me know if there's anything I can help with!

mmastrac commented:
If anyone wants to adopt this PR as their own, they have my blessing. I have been hoping for a block of time where I can revisit my outstanding PRs, but it hasn't appeared. 😞

JosJuice commented:
Here's my review of `ScheduleReads`, which was the only part I didn't review before.

---

Review status: 1 of 4 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1461 at r6](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KcDXD28w-F4WjrJeCu7:-KcDXD28w-F4WjrJeCu8:b-8v7e5c) ([raw file](https://github.com/dolphin-emu/dolphin/blob/141bc9288c21bd52e43ab1c65f6ff9b9d7c276dc/Source/Core/Core/HW/DVDInterface.cpp#L1461)):*
> ```cpp
>       if (rounded_offset != head_position)
>       {
>         // Unbuffered seek+read
> ```

There can probably be cases where it would be faster to wait for the head to continue as usual and read more data into a buffer that is about to become complete soon instead of seeking, considering that a seek always takes at least 45 ms according to our model. The old code always picked the faster of the two.

---

*[Source/Core/Core/HW/DVDInterface.cpp, line 1462 at r6](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-KcDX_vzCR3ANKDGeqno:-KcDX_w-n6QOIh8TViaq:b4spxch) ([raw file](https://github.com/dolphin-emu/dolphin/blob/141bc9288c21bd52e43ab1c65f6ff9b9d7c276dc/Source/Core/Core/HW/DVDInterface.cpp#L1462)):*
> ```cpp
>       {
>         // Unbuffered seek+read
>         ticks_until_completion += CalculateSeekTime(head_position, rounded_offset);
> ```

You're only adding the seek time, not the read time. It looks like a mistake, but could it be intentional because the read time for one block is baked into the value returned by CalculateSeekTime?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-:-KcDjrn_Z4j69_H2lxFT:b-n1gn3a)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Oh right, Reviewable lets you mark files as reviewed. I forgot.

---

Reviewed 2 of 4 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3701#-:-KcDqXG48qLlHg2_bbd_:bb5zsls)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
I've made a rebased PR: PR #4829

mmastrac commented:
Closing in favour of #4829
